lib/sort_by_length.rb

```diff
@@ -1,7 +1,40 @@
 # A method which will return an array of the words in the string
 #  sorted by the length of the word.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: I don't know! :(
+# Space complexity: O(n)
 def sort_by_length(my_sentence)
-  raise NotImplementedError, "Method not implemented"
+  array = my_sentence.split
+  words_by_length = {}
+  i = 0
+  while i < array.length #O(m)
+    if words_by_length[array[i].length] == nil
+      words_by_length[array[i].length] = []
+    end
+    words_by_length[array[i].length] << array[i]
+    i += 1
+  end
+
+  j = 0
+  new_array = [] #space = O(n)
+  while j < my_sentence.length #O(n)
+    if words_by_length[j] != nil
+      new_array << words_by_length[j]
+    end
```

Comment:
Above you've built a hash of lengths of words, and then in this loop  you put all the values from that array into `new_array`.  Then you essentially flatten new array into the result.

Why not do?

```ruby
(0..words_by_length.keys.max).each do |length| # O(n) where n is the number of words
  unless words_by_length[length].nil?
    words_by_length[length].each do |word|
      result << word
    end
  end
end
```

What you have is actually pretty close to bucket sort.  Neat.